hcl2template/addrs/plugin.go

```diff
@@ -98,67 +134,63 @@ func IsPluginPartNormalized(str string) (bool, error) {
 //
 // The following are valid source string formats:
 //
-//	name
 //	namespace/name
 //	hostname/namespace/name
 func ParsePluginSourceString(str string) (*Plugin, hcl.Diagnostics) {
-	ret := &Plugin{
-		Hostname:  "",
-		Namespace: "",
-	}
 	var diags hcl.Diagnostics
 
-	// split the source string into individual components
-	parts := strings.Split(str, "/")
-	if len(parts) != 3 {
-		diags = diags.Append(&hcl.Diagnostic{
+	// Trim both leading and trailing slashes for convenience
+	str = strings.Trim(str, "/")
+
+	url, err := url.Parse(str)
+	if err != nil {
+		return nil, diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
-			Summary:  "Invalid plugin source string",
-			Detail:   `The "source" attribute must be in the format "hostname/namespace/name"`,
+			Summary:  "Invalid source URL",
+			Detail:   "A source must be a valid URL",
 		})
-		return nil, diags
 	}
 
-	// check for an invalid empty string in any part
-	for i := range parts {
-		if parts[i] == "" {
-			diags = diags.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid plugin source string",
-				Detail:   `The "source" attribute must be in the format "hostname/namespace/name"`,
-			})
-			return nil, diags
-		}
+	if url.Scheme != "" {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid source URL",
+			Detail:   "A source must not contain a scheme",
+		})
 	}
 
-	// check the 'name' portion, which is always the last part
-	givenName := parts[len(parts)-1]
-	name, err := ParsePluginPart(givenName)
-	if err != nil {
-		diags = diags.Append(&hcl.Diagnostic{
+	if url.RawQuery != "" || url.RawFragment != "" {
+		return nil, diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
-			Summary:  "Invalid plugin type",
-			Detail:   fmt.Sprintf(`Invalid plugin type %q in source %q: %s"`, givenName, str, err),
+			Summary:  "Invalid source URL",
+			Detail:   "A source must not contain a query or a fragment",
+		})
+	}
+
+	str = url.String()
+
+	slashCount := strings.Count(str, "/")
+	if slashCount == 0 {
```

Comment:
I feel like this check should happen higher up but as I understand you are validating the url first then checking of paths, thats why it is here. You can achieve the same results using strings.Contains if slashCount is not used anywhere else.